### PR TITLE
feat: ogmios chain sync

### PR DESCRIPTION
**Pull request by mkazlauskas**
# Context

Projection service needs to have event source

# Proposed Solution

See commit log

### CODE REVIEW

File: packages/projection/src/chain-sync/ogmios/getNetworkInfo.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import { ogmiosToCore } from '@cardano-sdk/ogmios';
+
+export const getNetworkInfo = async (connectionConfig: ConnectionConfig): Promise<WithNetworkInfo> => {
+  let done = false;
+  const client = await createStateQueryClient(
+    await createInteractionContext(
+      (error) => {
+        throw error;
+      },
+      () => {
```
Comment (rhyslbw) [inline packages/projection/src/chain-sync/ogmios/getNetworkInfo.ts:13]:
``` ts
(code, reason) => {
  if (code === 1000) {
    // expected
 } else {
   throw new Error('Ogmios connection unexpectedly closed');
}
```

File: packages/projection/src/chain-sync/ogmios/getNetworkInfo.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
+// Tested in packages/e2e/test/ogmios
+import { ConnectionConfig, createInteractionContext, createStateQueryClient } from '@cardano-ogmios/client';
+import { WithNetworkInfo } from '../../operators';
+import { ogmiosToCore } from '@cardano-sdk/ogmios';
+
+export const getNetworkInfo = async (connectionConfig: ConnectionConfig): Promise<WithNetworkInfo> => {
+  let done = false;
+  const client = await createStateQueryClient(
```
Comment (rhyslbw) [inline packages/projection/src/chain-sync/ogmios/getNetworkInfo.ts:8]: Any reason to not use the [CardanoNode](https://github.com/input-output-hk/cardano-js-sdk/blob/e910b2310320418386f692f590d34416ee53445f/packages/core/src/CardanoNode/types/CardanoNode.ts#L33-L55) interface?

File: packages/projection/src/chain-sync/ogmios/createObservableChainSyncClient.ts [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+              point: ogmiosToCorePointOrOrigin(point),
+              requestNext,
+              tip: ogmiosToCoreTipOrOrigin(tip)
+            });
+          },
+          async rollForward({ block, tip }, requestNext) {
+            const coreBlock = ogmiosToCore.block(block);
+            if (!coreBlock) {
+              // Assuming it's an EBB
+              return;
```
[the remark this answers is not in the document]
Comment (mkazlauskas) [reply packages/projection/src/chain-sync/ogmios/createObservableChainSyncClient.ts]: Nice catch https://github.com/input-output-hk/cardano-js-sdk/pull/593/files#diff-1f0f1242eb1dbaba9c7185dba294abaa9e590f413e370c0dac74dd64a063d4bbR41

File: packages/projection/src/chain-sync/ogmios/createObservableChainSyncClient.ts [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    };
+  }),
+  intersection: ogmiosToCoreIntersection(intersection)
+});
+
+export const createObservableChainSyncClient =
+  (connectionConfig: ConnectionConfig): ObservableChainSyncClient =>
+  ({ points }) =>
+    new Observable((subscriber) => {
+      createSubscriberInteractionContext(connectionConfig, subscriber, 'OneTime')
```
[the remark this answers is not in the document]
Comment (mkazlauskas) [reply packages/projection/src/chain-sync/ogmios/createObservableChainSyncClient.ts]: Now shares a single connection 5cde0483e1d6f28c7f6d9d5d582dd45094b28b47

File: packages/projection/src/chain-sync/ogmios/getNetworkInfo.ts (hunk 1 above)
Comment (mkazlauskas) [reply packages/projection/src/chain-sync/ogmios/getNetworkInfo.ts]: Thanks! https://github.com/input-output-hk/cardano-js-sdk/pull/593/files#diff-d37d215964cf45049fba9ad64ec4cbe222235c119e6890b805c781833edff19cR66

File: packages/projection/src/chain-sync/ogmios/getNetworkInfo.ts (hunk 2 above)
Comment (mkazlauskas) [reply packages/projection/src/chain-sync/ogmios/getNetworkInfo.ts]: This is gone now, using [ObservableCardanoNode](https://github.com/input-output-hk/cardano-js-sdk/pull/593/files#diff-2bc9b1dc745118b9f203c03579483b6437a02054833d620145bb14830b9c1fdcR46) in place of this

File: packages/ogmios/src/CardanoNode/OgmiosObservableCardanoNode/createObservableInteractionContext.ts [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+              );
+            },
+            (code, reason) => {
+              if (code === 1000) {
+                logger.debug('Websocket closed with code 1000 (Normal Closure)');
+                subscriber.complete();
+              } else {
+                const message = `Websocket unexpectedly closed with code ${code}: ${reason}`;
+                logger.error(message);
+                // TODO: this error doesn't contain any details {code,reason}, maybe create a new error type?
```
Comment (rhyslbw) [inline packages/ogmios/src/CardanoNode/OgmiosObservableCardanoNode/createObservableInteractionContext.ts:72]: Sounds reasonable to define a custom error

File: packages/projection/src/projectIntoSink.ts [hunk 6]
```diff
@@ -121,32 +124,55 @@ export const projectIntoSink = <P extends object, PS extends P>(
...
+          if (blocks[0] !== 'origin' && initialChainSync.intersection.point.hash !== blocks[0].header.hash) {
+            // rollback to intersection, then sync from intersection
+            return rollbackAndSyncFromIntersection(initialChainSync, blocks[1]);
+          }
+          // intersection is at tip$ - no rollback, just sync from intersection
+          return syncFromIntersection(initialChainSync);
+        })
+      );
+    }),
+    // tap((evt) => {
```
Comment (rhyslbw) [inline packages/projection/src/projectIntoSink.ts:161]: What's this?

**Review by rhyslbw — COMMENTED** [review]
Great work @mkazlauskas :rocket:

File: packages/ogmios/src/CardanoNode/OgmiosObservableCardanoNode/createObservableInteractionContext.ts (hunk 5 above)
Comment (mkazlauskas) [reply packages/ogmios/src/CardanoNode/OgmiosObservableCardanoNode/createObservableInteractionContext.ts:72]: Agree, but we should be using CardanoNodeErrors in this context, and all of the errors used are currently Ogmios. If we add a new type, then we need to make sure to map ogmios ConnectionError into the new error everywhere else. If we don't then we end up with 2 connection error types that represent the same thing. Maybe let's tackle this separately?

Comment (rhyslbw) [reply packages/ogmios/src/CardanoNode/OgmiosObservableCardanoNode/createObservableInteractionContext.ts:72]: Yes, separate is best.

File: packages/projection/src/projectIntoSink.ts (hunk 6 above)
Comment (mkazlauskas) [reply packages/projection/src/projectIntoSink.ts]: Trash, removed it in 9f54088fb256f79e8da2c838b1244e92618f25b2 (squashed to appropriate commit)

File: packages/ogmios/src/CardanoNode/OgmiosObservableCardanoNode/createObservableInteractionContext.ts (hunk 5 above)
Comment (mkazlauskas) [reply packages/ogmios/src/CardanoNode/OgmiosObservableCardanoNode/createObservableInteractionContext.ts]: ADP-2734
[Updated the comment](https://github.com/input-output-hk/cardano-js-sdk/pull/593/commits/b1c7785c2b7cd554dc1adf9cf90db5897e4eaebc#diff-d37d215964cf45049fba9ad64ec4cbe222235c119e6890b805c781833edff19cR72-R73) in appropriate commit

**Review by rhyslbw — APPROVED** [review]

**Review by mirceahasegan — APPROVED** [review]
Out of this world 👽

**Pull request merged**